pkg/util/process/collector.go

```diff
@@ -0,0 +1,133 @@
+package process
+
+import (
+	"bufio"
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// DefaultProcMountPoint is the common mount point of the proc filesystem.
+	DefaultProcMountPoint = "/proc"
+)
+
+type processCollector struct {
+	pid            int
+	procMountPoint string
+	supported      bool
+
+	// Metrics.
+	currMaps *prometheus.Desc
+	maxMaps  *prometheus.Desc
+}
+
+// NewProcessCollector makes a new custom process collector used to collect process metrics the
+// default instrumentation doesn't support.
+func NewProcessCollector() prometheus.Collector {
+	return newProcessCollector(os.Getpid(), DefaultProcMountPoint)
+}
+
+func newProcessCollector(pid int, procMountPoint string) prometheus.Collector {
+	c := &processCollector{
+		pid:            pid,
+		procMountPoint: procMountPoint,
+		supported:      isSupported(procMountPoint),
+		currMaps: prometheus.NewDesc(
+			"process_memory_map_areas",
+			"Number of memory map areas allocated by the process.",
+			nil, nil,
+		),
+		maxMaps: prometheus.NewDesc(
+			"process_memory_map_areas_limit",
+			"Maximum number of memory map ares the process can allocate.",
+			nil, nil,
+		),
+	}
+
+	return c
+}
+
+// Describe returns all descriptions of the collector.
+func (c *processCollector) Describe(ch chan<- *prometheus.Desc) {
+	if !c.supported {
+		return
+	}
+
+	ch <- c.currMaps
+	ch <- c.maxMaps
+}
+
+// Collect returns the current state of all metrics of the collector.
+func (c *processCollector) Collect(ch chan<- prometheus.Metric) {
+	if !c.supported {
+		return
+	}
+
+	if value, err := c.getMapsCount(); err == nil {
```

Comment:
If err is not nil, we should return -1. By skipping this metric on error, we break metric collection. Same is true below.

Reply:
I asked Beorn cause I had a doubt whether this is true or not:
> It is OK to return a descriptor for which you never (or not always) return a metric.

Basically it's not a problem here. I think we've seen the issue you describe long time ago when working with the metrics aggregator helper in Cortex.